Light by light modifier 1100 pre1

rchudasa opened the pull request:
#### PR description:
This PR is submitted to reconstruct low pT photons down to 2 GeV for light by light scattering. It is just about relaxing few thresholds in python config with no changes in c++ source code.  A modifier has been added to access the low pt photon reconstruction. Photon and electron efficiency plots and timing with default and modified reconstruction can be found in slides [1] presented in Reconstruction and Analysis Tools meeting. The modified reconstruction has been reviewed by Egamma POG convener @Sam-Harper .
[1]https://indico.cern.ch/event/836517/contributions/3513898/attachments/1887027/3111085/cms_reco_meeting_26thJuly_2019.pdf

Sam-Harper committed:
adding light by light modifier

rchudasa committed:
updating rechits and ecal digis selectors

rchudasa committed:
Modifier for lbyl added to Eras repo

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27633/11126

- This PR adds an extra 44KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/StandardSequences/python/Eras.py  modified in PR(s): #27555


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27633/11126/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27633/11126/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

slava77 commented on Configuration/Eras/python/Modifier_lightByLightLowPt_cff.py:
```diff
@@ -0,0 +1,4 @@
+import FWCore.ParameterSet.Config as cms
```
Configuration/ProcessModifiers/python might be a better place

slava77 commented:
> Photon and electron efficiency plots and timing with default and modified reconstruction can be found in slides [1] presented in Reconstruction and Analysis Tools meeting. The modified reconstruction has been reviewed by Egamma POG convener @Sam-Harper .
> [1]https://indico.cern.ch/event/836517/contributions/3513898/attachments/1887027/3111085/cms_reco_meeting_26thJuly_2019.pdf

I missed this part of the meeting.
What do you imply by the "default reco"?
Naively, given the 10_3_X context and HI PDs, I can guess that this is about pp_on_AA,
but I'm not sure.

Is this expected to be a once-in-CMS-lifetime for 10_3_X use case or should this be a supported workflow?

rchudasa commented:
> > Photon and electron efficiency plots and timing with default and modified reconstruction can be found in slides [1] presented in Reconstruction and Analysis Tools meeting. The modified reconstruction has been reviewed by Egamma POG convener @Sam-Harper .
> > [1]https://indico.cern.ch/event/836517/contributions/3513898/attachments/1887027/3111085/cms_reco_meeting_26thJuly_2019.pdf
> 
> I missed this part of the meeting.
> What do you imply by the "default reco"?
> Naively, given the 10_3_X context and HI PDs, I can guess that this is about pp_on_AA,
> but I'm not sure.
> 
> Is this expected to be a once-in-CMS-lifetime for 10_3_X use case or should this be a supported workflow?

The default reco is pp reconstruction with era "Run2_2018" and on top of that light by light modifier has been added. We don't use or modify pp_on_AA era as it modifies many Egamma thresholds to higher values which we don't want. 
This is a supported workflow for higher CMSSW release and even for pp analyses, if required. :)

slava77 commented:
> The default reco is pp reconstruction with era "Run2_2018" 

uhm, so this make a stronger case for backporting this at best to 10_6_X to be based on the UL release.

rchudasa committed:
interesting rechit threshold updated

slava77 commented:
@rchudasa 
there is a merge conflict in this PR now
Please update.

rchudasa committed:
ecalDrivenElectronseed merge conflict solved

cmsbuild commented:
The code-checks are being triggered in jenkins.

rchudasa reviewed (commented):
> @rchudasa
> there is a merge conflict in this PR now
> Please update.

@slava77 , I just committed a change now. Please note that I have not changed the location of lightbylight modifier to ProcessModifier repo yet, as cmsDriver command could not identify that.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27633/11237

- This PR adds an extra 44KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/StandardSequences/python/Eras.py  modified in PR(s): #27665

cmsbuild commented:
A new Pull Request was created by  @rchudasa  for master.

It involves the following packages:

Configuration/Eras
Configuration/StandardSequences
RecoEcal/EgammaClusterProducers
RecoEgamma/EgammaElectronProducers
RecoEgamma/EgammaIsolationAlgos
RecoEgamma/EgammaPhotonProducers
RecoParticleFlow/PFProducer
RecoParticleFlow/PFTracking


@perrotta, @cmsbuild, @franzoni, @slava77, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@varuns23, @mmarionncern, @Sam-Harper, @argiro, @makortel, @felicepantaleo, @jainshilpi, @cbernet, @GiacomoSguazzoni, @rovere, @lgray, @Martin-Grunewald, @seemasharmafnal, @hatakeyamak, @VinInn, @dgulhan, @bachtis this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
On 8/1/19 8:34 AM, rchudasa wrote:
> Please note that I have not changed the location of lightbylight
> modifier to ProcessModifier repo yet, as cmsDriver command could not
> identify that.
> 
> ------------------------------------------------------------------------

it should be passed with --procModifiers options

slava77 commented:
>  I tried to keep this modifier in ProcessModifiers repo as suggested by you and tried to run cmsDriver command with --era Run2_2018 --procModifiers lightByLightLowPt but it tries to find this Modifier in Eras repo, could you please suggest what else should I modify, so that cmsDriver command recognizes it from ProcessModifiers repo?

This sounds strange. By construction ```--procModifiers``` looks in Configuration.ProcessModifiers.
https://github.com/cms-sw/cmssw/blob/master/Configuration/Applications/python/ConfigBuilder.py#L2092
Note that you should drop ```Modifier_``` prefix from the file name.

rchudasa commented:
> > I tried to keep this modifier in ProcessModifiers repo as suggested by you and tried to run cmsDriver command with --era Run2_2018 --procModifiers lightByLightLowPt but it tries to find this Modifier in Eras repo, could you please suggest what else should I modify, so that cmsDriver command recognizes it from ProcessModifiers repo?
> 
> This sounds strange. By construction `--procModifiers` looks in Configuration.ProcessModifiers.
> https://github.com/cms-sw/cmssw/blob/master/Configuration/Applications/python/ConfigBuilder.py#L2092
> Note that you should drop `Modifier_` prefix from the file name.

Yes, I dropped Modifier_ from the name but the problem is, the cmsDriver command tries to search in Eras repository, although I gave --procModifiers option. Here's the error: CMSSW_11_0_X_2019-07-29-2300/python/Configuration/StandardSequences/Eras.py", line 81, in __init__
    eObj=getattr(__import__('Configuration.Eras.Modifier_'+e+'_cff',globals(),locals(),[e],0),e)
ImportError: No module named Modifier_lightByLightLowPt_cff

makortel commented:
> Yes, I dropped Modifier_ from the name but the problem is, the cmsDriver command tries to search in Eras repository, although I gave --procModifiers option. Here's the error: CMSSW_11_0_X_2019-07-29-2300/python/Configuration/StandardSequences/Eras.py", line 81, in init
> eObj=getattr(import('Configuration.Eras.Modifier_'+e+'_cff',globals(),locals(),[e],0),e)
> ImportError: No module named Modifier_lightByLightLowPt_cff
> 

You need to remove your additions  to `Eras.py`, namely
https://github.com/cms-sw/cmssw/blob/645b6198df584667c0b6265af657135285c26f1f/Configuration/StandardSequences/python/Eras.py#L70

slava77 commented:
about the naming of the modifier, I'd like to propose some decoration that refers to a POG or a PAG. One somewhat clear option is ```egamma_photon_lowPt```.
The specific goal of the proposed change is really to capture the (semi)exclusive diphoton production. Perhaps in the HI case gamma-gamma dominates, but at lower Z the diffractive component can be large as well.

OTOH, if the longer term goal for this is to include coverage for other final states, perhaps a less specific name is more practical like ```exclusive_lowPt```

rchudasa commented:
> about the naming of the modifier, I'd like to propose some decoration that refers to a POG or a PAG. One somewhat clear option is `egamma_photon_lowPt`.
> The specific goal of the proposed change is really to capture the (semi)exclusive diphoton production. Perhaps in the HI case gamma-gamma dominates, but at lower Z the diffractive component can be large as well.
> 
> OTOH, if the longer term goal for this is to include coverage for other final states, perhaps a less specific name is more practical like `exclusive_lowPt`

@slava77 , currently we don't have any plan to extend this to any other final state and the current reconstruction modifies Photons and electrons both and it's more for exclusive process, we would like to propose 'egamma_lowPt_exclusive'. Is it ok with you?

slava77 commented:
On 8/2/19 9:24 AM, rchudasa wrote:
> we would like to propose 'egamma_lowPt_exclusive'. Is it ok with you?

sounds fine to me

fabiocos commented:
@rchudasa I agree with comments by @slava77 both about the modifier location (we are already abusing of Era, and a cleaning would be advisable https://github.com/cms-sw/cmssw/issues/22718 ) and the name, I would prefer to avoid overspecializing the name of top level configurations

rchudasa committed:
Light by light modifier renamed to egamma_lowPt_exclusive and repo changed to process modifier

cmsbuild commented:
The code-checks are being triggered in jenkins.

rchudasa committed:
Light by light modifier renamed to egamma_lowPt_exclusive in corresponding Egamma scripts

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27633/11295

- This PR adds an extra 48KB to repository

- Found files with invalid states:
  - Configuration/Eras/python/Modifier_lightByLightLowPt_cff.py:
    - Added: ce82fec478d796f111a1d42003d38cd8406b4e62
    - Deleted: e7abd482fe9af0490b6840687ac9d4e2168efef0


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/StandardSequences/python/Eras.py  modified in PR(s): #27665, #27672

cmsbuild commented:
Pull request #27633 was updated. @perrotta, @cmsbuild, @franzoni, @slava77, @fabiocos, @davidlange6 can you please check and sign again.

rchudasa commented:
@slava77 @fabiocos @makortel : Thank you for your suggestions. I have renamed the light by light modifier to egamma_lowPt_exclusive and also kept it in ProcessModifiers repository.

fabiocos commented:
please test

not really testing the new feature here

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 287747f8783845804c3950e452abffee9333fce4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bda038/1836/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bda038/1836/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-bda038/11634.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 2715989
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2715659
 - DQMHistoTests: Total skipped: 329
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

slava77 commented:
@rchudasa @echapon 
please post a cmsDriver command which can be used to test the proposed update together with an appropriate input file (preferably readable via xrootd).
Thank you.

slava77 commented on RecoEgamma/EgammaPhotonProducers/python/photons_cfi.py:
```diff
@@ -150,3 +150,11 @@
+                           minSCEtEndcap = 1.0) #default 10
+egamma_lowPt_exclusive.toModify(islandPhotons,
+                           minSCEtBarrel = 1.0, #default 5
+                           minSCEtEndcap = 1.0) #default 15
```
this is not a part of the regular pp reco.
Why is this modified?

rchudasa replied:
ok, we will not change the island Photon thresholds (updating it in the next commit).

slava77 commented on RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py:
```diff
@@ -209,3 +209,9 @@ def _findIndicesByModule(name):
 )
+
+from Configuration.ProcessModifiers.egamma_lowPt_exclusive_cff import egamma_lowPt_exclusive
+egamma_lowPt_exclusive.toModify(particleFlowBlock.elementImporters[2],
```
please use ```_findIndicesByModule``` instead of direct indexing.
A less preferred solution would be to move this change up, to be right after the definition of particleFlowBlock, before all other modifiers are called

rchudasa replied:
could you please suggest how exactly should I do it?

slava77 replied:
visually, it looks like  replacing 2 with ```_findIndicesByModule("SuperClusterImporter")[0]``` should work,
perhaps more effective is to do
```python
for imp in particleFlowBlock.elementImporters:
  if imp..importerName.value() == "SuperClusterImporter":
    _scImporter = imp
egamma_lowPt_exclusive.toModify(_scImporter, ...
```

rchudasa commented:
> @rchudasa @echapon
> please post a cmsDriver command which can be used to test the proposed update together with an appropriate input file (preferably readable via xrootd).
> Thank you.
@slava77 @echapon 
Here's the cmsDriver command with xrootd input 

cmsDriver.py step3 --mc --eventcontent AODSIM --datatier AODSIM --conditions 110X_upgrade2018_realistic_Candidate_2019_08_03_11_37_17 --step RAW2DIGI,L1Reco,RECO --geometry DB:Extended -n -1 --era Run2_2018 --procModifiers egamma_lowPt_exclusive --customise Configuration/DataProcessing/Utils.addMonitoring --filein=root://cmsxrootd.fnal.gov//store/group/phys_diffraction/lbyl_2018/for_pr/FlatGamma_pt_digi_raw_11X.root --customise_commands 'process.AODSIMoutput.outputCommands.extend(["keep *_towerMaker_*_*", "keep *_siPixelRecHits_*_*", "keep *_siPixelClusters_*_*", "keep *_interestingGedEgammaIsoESDetId_*_*", "keep *_interestingGedEgammaIsoHCALDetId_*_*", "keep *_interestingGedEleIsoDetIdEB_*_*", "keep *_interestingGedEleIsoDetIdEE_*_*", "keep *_interestingGedGamIsoDetIdEB_*_*","keep *_interestingGedGamIsoDetIdEE_*_*", "keep EBDigiCollection_ecalDigis_*_*","keep EEDigiCollection_ecalDigis_*_*","keep ESDigiCollection_ecalPreshowerDigis_*_*","keep ZDCRecHitsSorted_zdcreco_*_*"])'

slava77 commented:
@rchudasa 
Thank you for the cmsDriver command.
if you need modifications to the output AODSIM, it is more appropriate to make them with egamma_lowPt_exclusive in the code base. Please add them to this PR.
Please also escape the command in full with \`\`\` on both sides so that it's displayed appropriately (https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519177827 looks like missing some underscores and stars due to confusion with the markdown special characters)

Please update your tests (extending what was shown in the reco meeting) with details of the output file sizes. I expect that the extra data in your ```keep ``` statements increases the AOD size by a factor of a few.

slava77 commented:
how was the file FlatGamma_pt_digi_raw_11X.root (proposed for testing)  made?

slava77 commented:
> how was the file FlatGamma_pt_digi_raw_11X.root (proposed for testing) made?

Looking at the provenance, I see that this is a photon gun MC sample with photon in |eta|<3 and pt from 1 to 10 GeV.
I think that this is insufficient for technical performance tests.
Please also provide a data file.

rchudasa commented:
> > how was the file FlatGamma_pt_digi_raw_11X.root (proposed for testing) made?
> 
> Looking at the provenance, I see that this is a photon gun MC sample with photon in |eta|<3 and pt from 1 to 10 GeV.
> I think that this is insufficient for technical performance tests.
> Please also provide a data file.
Yes, the digi raw file was produced with flat pt photon sample produced for pt: 1-20 GeV within |eta|<3. The dtata file can be found here: /eos/cms/store/group/phys_diffraction/lbyl_2018/F9B1A876-3F7D-8449-8437-894248B60877.root. This data file contains good lumi sections.

slava77 commented:
> The dtata file can be found here: /eos/cms/store/group/phys_diffraction/lbyl_2018/F9B1A876-3F7D-8449-8437-894248B60877.root. This data file contains good lumi sections.

Which GT should be used with this data file? Is the run2_data OK or do we need run2_data_promptlike_hi ?
Please clarify, thank you.

slava77 commented:
> Which GT should be used with this data file? Is the run2_data OK or do we need run2_data_promptlike_hi ?

I see that it runs with auto:run2_data. perhaps this is good enough for the tests.

slava77 commented:
> Please update your tests (extending what was shown in the reco meeting) with details of the output file sizes. I expect that the extra data in your `keep ` statements increases the AOD size by a factor of a few.

I should have been looking at the slides more carefully.
There is a "Size:" bullet for both cases.
However, please clarify if the default reco size was measured with the keep statements or using the default AOD definition.

slava77 commented:
about timing: I was not able to reproduce the significant increase in CPU time between the default and the modified recos (you showed the avg go up by 40% from 0.45 s and the max go up by more than 2 from 881 s); in my test there is about 5% increase in the CPU time.
@rchudasa   Please clarify where the tests were made and also how certain are you that nothing else slowed down the job in the modified reco case.

The only relevant CPU cost increases I see are in 
- ecalDrivenElectronSeeds : \~+1% of the reco job
- allConversions : \~+2% of the reco job
- some other egamma modules with significant self-increase add up to another 1-2% of the reco job.

The (extreme) tails in the CPU processing are due to modules not really related to the changes in ```egamma_lowPt_exclusive``` modifier, primarily from the tracking modules. This means that some tracking activity veto is important for the production setup.


using the same file F9B1A876-3F7D-8449-8437-894248B60877.root and running on 10K events I see slightly different results in 11_0_0_pre5
- default AOD size is 19 kiB/event, compared to 34 kB shown in the reco meeting. 
    - I'm guessing that these 34 kB were obtained with the non-standard event content.  Please check and confirm.
- light-by-light setup with the keep statements is 31 kiB/event, which is closer to the 36 kB shown in the reco meeting.

slava77 commented:
I ran a reco job for all events in the file and there is at least one (out of 70K) event with about 15% related to the egamma_lowPt configuration updates, compared to 8% increase on average. 

In my test the average time per event with this PR was 1.39 s with a max of 2699 s. Here 
- 98.7% of events completed with time under 4 s/ev and an average in this range of 0.85 s/event. The remaining 2.3% of events with time > 4 s/ev had an average processing time of 43 s/event.
- 99.5% of events completed with time under 10 s/ev and an average in this range of 0.91 s/event. The remaining 0.5% of events with time > 10 s/ev had an average processing time of 103 s/event.

This is a somewhat clear indication of fairly long tails at sub-% level. IMO, these better be filtered out to avoid stuck processing or/and RSS memory over-use which could lead to occasional crashes. 

For this PR the main missing piece now is the event content definition in the code base with the modifier.
Filtering can be addressed separately.

rchudasa commented:
> about timing: I was not able to reproduce the significant increase in CPU time between the default and the modified recos (you showed the avg go up by 40% from 0.45 s and the max go up by more than 2 from 881 s); in my test there is about 5% increase in the CPU time.
> @rchudasa Please clarify where the tests were made and also how certain are you that nothing else slowed down the job in the modified reco case.
> 
> The only relevant CPU cost increases I see are in
> 
>     * ecalDrivenElectronSeeds : ~+1% of the reco job
> 
>     * allConversions : ~+2% of the reco job
> 
>     * some other egamma modules with significant self-increase add up to another 1-2% of the reco job.
> 
> 
> The (extreme) tails in the CPU processing are due to modules not really related to the changes in `egamma_lowPt_exclusive` modifier, primarily from the tracking modules. This means that some tracking activity veto is important for the production setup.
> 
> using the same file F9B1A876-3F7D-8449-8437-894248B60877.root and running on 10K events I see slightly different results in 11_0_0_pre5
> 
>     * default AOD size is 19 kiB/event, compared to 34 kB shown in the reco meeting.
>       
>       * I'm guessing that these 34 kB were obtained with the non-standard event content.  Please check and confirm.
> 
>     * light-by-light setup with the keep statements is 31 kiB/event, which is closer to the 36 kB shown in the reco meeting.

@slava77 
I re-checked the timing with the default and modified reconstruction for 10K events. I tried to run in CMSSW_10_3_2 and CMSSW_11_0_X_2019-07-29-2300. I used 103X_dataRun2_Prompt_fixEcalADCToGeV_v1 global tag for CMSSW_10_3 and auto:data_run2 for 11_X version.  I am able to get the same results as I presented in my slides for CMSSW_10_3_2 and your results for CMSSW_11_0_X. 

Default reconstruction: 
* CMSSW_10_3_2: Average time: 0.48 s, Max time: 969 s
* CMSSW_11_0_X_2019-07-29-2300: Average time: 0.528 s, Max time: 828 s

Modified reconstruction:
* CMSSW_10_3_2: Average time: 0.64 s, Max time: 2087 s
* CMSSW_11_0_X_2019-07-29-2300: Average time: 0.53 s, Max time: 927 s

and you are right about the event size, default AOD content size is 19kB/event. 
I add the event content definition in code.

rchudasa committed:
Event content code based, island photon threshold not changed and Slava's comment on PFBlock addressed

cmsbuild commented:
The code-checks are being triggered in jenkins.

rchudasa committed:
removed commented lines in particleFlowBlock

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27633/11407

- This PR adds an extra 96KB to repository

- Found files with invalid states:
  - Configuration/Eras/python/Modifier_lightByLightLowPt_cff.py:
    - Added: ce82fec478d796f111a1d42003d38cd8406b4e62
    - Deleted: e7abd482fe9af0490b6840687ac9d4e2168efef0

cmsbuild commented:
Pull request #27633 was updated. @perrotta, @cmsbuild, @franzoni, @slava77, @fabiocos, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: fb097ff535b1d3f4f036c184e8cebea4ce87941b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bda038/1975/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bda038/1975/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939166
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 142 log files, 14 edm output root files, 34 DQM output files

slava77 commented:
@rchudasa 
compared to the test command in https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519177827
I'm missing the event content modifications for the following in the latest version of the PR
```
"keep *_interestingGedEgammaIsoESDetId_*_*", "keep *_interestingGedEgammaIsoHCALDetId_*_*", 
"keep *_interestingGedEleIsoDetIdEB_*_*", "keep *_interestingGedEleIsoDetIdEE_*_*",
"keep *_interestingGedGamIsoDetIdEB_*_*","keep *_interestingGedGamIsoDetIdEE_*_*",
"keep EBDigiCollection_ecalDigis__*","keep EEDigiCollection_ecalDigis__*","keep ESDigiCollection_ecalPreshowerDigis__*",
```
Please check and confirm that these are not needed or update the PR to include the missing parts.

Also, compared to the test setup, the following were added
```
     'keep ZDCDataFramesSorted_hcalDigis_*_*',
    'keep ZDCDataFramesSorted_castorDigis_*_*',
    'keep QIE10DataFrameHcalDataFrameContainer_hcalDigis_ZDC_*'
```
ZDC data is small, so it seems OK to add if it's needed.

rchudasa commented:
> @rchudasa
> compared to the test command in [#27633 (comment)](https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519177827)
> I'm missing the event content modifications for the following in the latest version of the PR
> 
> ```
> "keep *_interestingGedEgammaIsoESDetId_*_*", "keep *_interestingGedEgammaIsoHCALDetId_*_*", 
> "keep *_interestingGedEleIsoDetIdEB_*_*", "keep *_interestingGedEleIsoDetIdEE_*_*",
> "keep *_interestingGedGamIsoDetIdEB_*_*","keep *_interestingGedGamIsoDetIdEE_*_*",
> "keep EBDigiCollection_ecalDigis__*","keep EEDigiCollection_ecalDigis__*","keep ESDigiCollection_ecalPreshowerDigis__*",
> ```
> 
> Please check and confirm that these are not needed or update the PR to include the missing parts.
> 
> Also, compared to the test setup, the following were added
> 
> ```
>      'keep ZDCDataFramesSorted_hcalDigis_*_*',
>     'keep ZDCDataFramesSorted_castorDigis_*_*',
>     'keep QIE10DataFrameHcalDataFrameContainer_hcalDigis_ZDC_*'
> ```
> 
> ZDC data is small, so it seems OK to add if it's needed.

I consulted @Sam-Harper and concluded that we dont need those event contents as they are already present in reducedEgamma rechit collection. We require ZDC event content for the analysis.

slava77 commented:
> I consulted @Sam-Harper and concluded that we dont need those event contents as they are already present in reducedEgamma rechit collection. We require ZDC event content for the analysis.

Do you really mean "reducedEgamma"? This is a module from miniAOD.
In the earlier discussion we concluded that you will not run/need miniAOD.

Perhaps you actually mean other reduced rechit collections available in AOD.
Please clarify/confirm.

rchudasa commented:
> > I consulted @Sam-Harper and concluded that we dont need those event contents as they are already present in reducedEgamma rechit collection. We require ZDC event content for the analysis.
> 
> Do you really mean "reducedEgamma"? This is a module from miniAOD.
> In the earlier discussion we concluded that you will not run/need miniAOD.
> 
> Perhaps you actually mean other reduced rechit collections available in AOD.
> Please clarify/confirm.

Sorry, for the confusion. We will use AOD content and to be precise, reducedEcalRecHitsEB, reducedEcalRecHitsEE, reducedEcalRecHitsES collections.

slava77 commented:
+1

for  https://github.com/cms-sw/cmssw/pull/27633 fb097ff
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences (the new process modifier is not exercised)
- the latest updates should not change conclusions made in test in https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519759086 (or in the PR description). Some moderate cleanup of the keep statements  should reduce the event size by \~1 kB/event to just under 30 kB/evt (compared to an increase from 19 to 31 kiB observed in the earlier tests)

slava77 commented:
@fabiocos @kpedro88 
please check this and comment or merge some time soon.
Thank you.

kpedro88 commented:
+1
At some point (once decisions on production are made), a workflow using this modifier should be put into the matrix.

kpedro88 commented:
merge

fabiocos commented:
+operations

the addition of the modifier is correct, although it is not activated by default in a test workflow

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
